tests/share/search/test_indexing.py

```diff
@@ -0,0 +1,370 @@
+    ])
+    def message(self, request):
+        version, model, ids, payload = request.param
+
+        self.EXPECTED_IDS = ids
+        self.EXPECTED_MODEL = model
+        self.EXPECTED_VERSION = version
+
+        return kombu.Message(
+            content_type='application/json',
+            body=json.dumps(payload),
+        )
+
+    def test_wrap(self, message):
+        message = indexing.IndexableMessage.wrap(message)
+
+        assert list(message) == self.EXPECTED_IDS
+        assert message.model == self.EXPECTED_MODEL
+        assert message.protocol_version == self.EXPECTED_VERSION
+        assert isinstance(message, self.VERSION_MAP[self.EXPECTED_VERSION])
+
+    @pytest.mark.parametrize('version', [30, 10, None, 'Foo', '1', '0', {}, []])
+    def test_invalid_version(self, version):
+        with pytest.raises(ValueError) as e:
+            indexing.IndexableMessage.wrap(kombu.Message(
+                content_type='application/json',
+                body=json.dumps({'version': version}),
+            ))
+        assert e.value.args == ('Invalid version "{}"'.format(version), )
+
+
+class TestMessageFlattener:
+
+    def message_factory(self, model='CreativeWork', ids=None):
+        if ids is None:
+            ids = [1, 2, 3]
+
+        msg = indexing.IndexableMessage.wrap(kombu.Message(
+            content_type='application/json',
+            body=json.dumps({model: ids}),
+        ))
+
+        msg.message.ack = mock.Mock()
+        msg.message.requeue = mock.Mock()
+
+        return msg
+
+    @pytest.mark.parametrize('number', [0, 1, 5])
+    def test_len(self, number):
+        assert len(indexing.MessageFlattener([self.message_factory() for _ in range(number)])) == number * 3
+
+    def test_list(self):
+        assert list(indexing.MessageFlattener([])) == []
+
+        assert list(indexing.MessageFlattener([
+            self.message_factory(ids=[1, 2, 3]),
+        ])) == [1, 2, 3]
+
+        assert list(indexing.MessageFlattener([
+            self.message_factory(ids=[1, 2, 3]),
+            self.message_factory(ids=[4]),
+            self.message_factory(ids=[]),
+            self.message_factory(ids=[]),
+            self.message_factory(ids=[]),
+            self.message_factory(ids=[5, 6]),
+        ])) == [1, 2, 3, 4, 5, 6]
+
+    def test_moves_to_pending(self):
+        messages = [
+            self.message_factory(ids=[]),
+            self.message_factory(ids=[]),
+            self.message_factory(ids=[0]),
+            self.message_factory(ids=[1, 2]),
+            self.message_factory(ids=[3, 4, 5]),
+            self.message_factory(ids=[6, 7, 8, 9]),
+        ]
+
+        flattener = indexing.MessageFlattener(messages)
+        assert flattener.pending == []
+
+        # -1 as a static offset (starts with 2 1s)
+        # +2 to account for the 2 empty messages at the begining of the list
+        lengths = [i - 1 + 2 for i in range(2, 6) for _ in range(i)]
```

Comment:
I appreciate the logic, but this is way more opaque than just writing out a list of the expected lengths, and would require more thought to update if the test case ever changes. It also makes a longer list than necessary (extra 6s on the end), which could be confusing.
```
lengths = [3, 3, 4, 4, 4, 5, 5, 5, 5, 6]
```